chasm/lib/activity/activity.go

```diff
@@ -218,8 +214,58 @@ func (a *Activity) PopulateRecordActivityTaskStartedResponse(ctx chasm.Context,
 	return nil
 }
 
-func (a *Activity) RecordCompletion(_ chasm.MutableContext) error {
-	return serviceerror.NewUnimplemented("RecordCompletion is not implemented")
+func (a *Activity) RecordCompletion(ctx chasm.MutableContext, applyFn func(ctx chasm.MutableContext) error) error {
+	return applyFn(ctx)
+}
+
+func (a *Activity) recordActivityTimedOut(ctx chasm.MutableContext, timeoutType enumspb.TimeoutType) error {
+	outcome, err := a.Outcome.Get(ctx)
+	if err != nil {
+		return err
+	}
+
+	failure := &failurepb.Failure{
+		Message: fmt.Sprintf("activity %s timed out", timeoutType.String()),
+		FailureInfo: &failurepb.Failure_TimeoutFailureInfo{
+			TimeoutFailureInfo: &failurepb.TimeoutFailureInfo{
+				TimeoutType: timeoutType,
+			},
+		},
+	}
+
+	switch timeoutType {
+	// Schedule to start and schedule to close timeouts are not retried so we set the outcome failure directly and
+	// leave the attempt failure as is.
+	case enumspb.TIMEOUT_TYPE_SCHEDULE_TO_START,
+		enumspb.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE:
+		outcome.Variant = &activitypb.ActivityOutcome_Failed_{
+			Failed: &activitypb.ActivityOutcome_Failed{
+				Failure: failure,
+			},
+		}
+
+	// Start to close timeouts are retried attempts so we update the attempt failure info but leave the outcome failure
+	// empty to avoid duplication
+	case enumspb.TIMEOUT_TYPE_START_TO_CLOSE:
+		attempt, err := a.Attempt.Get(ctx)
+		if err != nil {
+			return err
+		}
+
+		currentTime := timestamppb.New(ctx.Now(a))
+
+		attempt.LastFailureDetails = &activitypb.ActivityAttemptState_LastFailureDetails{
+			LastFailure:     failure,
+			LastFailureTime: currentTime,
+		}
+		attempt.LastAttemptCompleteTime = currentTime
+		outcome.Variant = &activitypb.ActivityOutcome_Failed_{}
```

Comment:
The outcome shouldn't be updated here unless you know for sure that this is the last attempt.